src/array.vala

```diff
@@ -92,28 +91,69 @@ public class Array<T>: Object
         return p;
     }
 
-    public unowned T
-    get_scalar(ssize_t [] index)
+    public void*
+    get_pointer ([CCode (array_length = false)] ssize_t[] index)
     {
-        return (T) ((uint8*) _data.get_data () + _offset_for_index (index));
+        return (uint8*) _data.get_data () + _offset_for_index (index);
     }
 
-    public ArrayIterator<T> iterator()
+    public Value
+    get_value ([CCode (array_length = false)] ssize_t[] index)
     {
-        return new ArrayIterator<T>(this);
+        var _value = Value (scalar_type);
+        if (_value.fits_pointer ()) {
+            Memory.copy (_value.peek_pointer (), get_pointer (index), scalar_size);
+        } else if (scalar_type == Type.BOXED) {
+            _value.set_boxed (get_pointer (index));
+        } else if (scalar_type == typeof (string)) {
+            _value.set_string ((string) get_pointer (index));
```

Comment:
In our approach, GType is exclusive used for transformations and basic introspection.

For strings, we can use `instance_size` as the maximum length including null-terminator. We already have the guarantee that `sizeof (char) == 1`.

Beyond that we should not support other array-like scalars because it's a nasty hack.